sw/host/opentitanlib/src/test_utils/mem.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use arrayvec::ArrayVec;
+use std::time::Duration;
+
+use crate::io::uart::Uart;
+use crate::test_utils::e2e_command::TestCommand;
+use crate::test_utils::rpc::{UartRecv, UartSend};
+use crate::test_utils::status::Status;
+
+// Bring in the auto-generated sources.
+include!(env!("mem"));
+
+impl MemRead32Req {
+    pub fn execute(uart: &dyn Uart, address: u32) -> Result<u32> {
+        TestCommand::MemRead32.send(uart)?;
+        let op = MemRead32Req { address: address };
+        op.send(uart)?;
+        let resp = MemRead32Resp::recv(uart, Duration::from_secs(300), false)?;
+        Ok(resp.value)
+    }
+}
+
+impl MemReadReq {
+    pub fn execute(uart: &dyn Uart, start_address: u32, data: &mut [u8]) -> Result<()> {
+        let bytes_to_read = data.len();
+        let mut bytes_read = 0_usize;
+        while bytes_read < bytes_to_read {
+            let op_size = std::cmp::min(bytes_to_read - bytes_read, 256_usize);
+            TestCommand::MemRead.send(uart)?;
+            let op = MemReadReq {
+                address: start_address + (bytes_read as u32),
+                data_len: op_size.try_into().unwrap(),
+            };
+            op.send(uart)?;
+            let resp = MemReadResp::recv(uart, Duration::from_secs(300), false)?;
+            data[bytes_read..(bytes_read + op_size)].copy_from_slice(&resp.data.as_slice());
+            bytes_read += op_size;
+        }
+        Ok(())
+    }
+}
+
+impl MemWrite32Req {
+    pub fn execute(uart: &dyn Uart, address: u32, value: u32) -> Result<()> {
+        TestCommand::MemWrite32.send(uart)?;
+        let op = MemWrite32Req {
+            address: address,
+            value: value,
+        };
+        op.send(uart)?;
+        Status::recv(uart, Duration::from_secs(300), false)?;
+        Ok(())
+    }
+}
+
+impl MemWriteReq {
+    pub fn execute(uart: &dyn Uart, start_address: u32, data: &[u8]) -> Result<()> {
+        let bytes_to_write = data.len();
+        let mut bytes_written = 0_usize;
+        while bytes_written < bytes_to_write {
+            let op_size = std::cmp::min(bytes_to_write - bytes_written, 256_usize);
```

Comment:
It would be nice to avoid a constant and compute this number. I think it can be done using either `std::mem::size_of_val` but it would require to create a dummy value. Otherwise, you can go full rusty with this abomination:
```rust
struct MyStruct {
    foo: [u8; 256usize],
    // blabla
}

fn get_size_of_return_type<F, T, U>(_f: F) -> usize
where
    F: FnOnce(T) -> U
{
    std::mem::size_of::<U>()
}

fn main() {
    println!("size: {}", get_size_of_return_type(|x: MyStruct| x.foo));
}
```